### PR TITLE
Adding option to view full stepwise context after every run

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

This PR adds an expandable/collapsible section to every stepwise planner result in the prompt dialog to show more details about the thought process.

Details for other semantic dependencies will be added in a follow-up PR: https://github.com/microsoft/chat-copilot/issues/150

### Description

Screenshot 2023-08-11 at 10 34 40 AM

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
~~- [ ] All unit tests pass, and I have added new tests where possible~~
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: webapi/Models/Response/BotResponsePrompt.cs [hunk 1]
```diff
@@ -9,6 +9,29 @@ namespace CopilotChat.WebApi.Models.Response;
 /// </summary>
 public class BotResponsePrompt
 {
+    /// <summary>
+    /// Information about semantic dependencies of the prompt.
+    /// </summary>
+    public class SemanticDependency<T>
```
Comment (teresaqhoang) [inline webapi/Models/Response/BotResponsePrompt.cs:15]: Context for other semantic dependencies will be added in a separate PR: https://github.com/microsoft/chat-copilot/issues/150

File: webapi/Models/Response/StepwiseThoughtProcess.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    /// Steps taken execution stat.
+    /// </summary>
+    [JsonPropertyName("stepsTaken")]
+    public string StepsTaken { get; set; } = string.Empty;
+
+    /// <summary>
+    /// Time taken to fulfil the goal.
+    /// </summary>
+    [JsonPropertyName("timeTaken")]
+    public string TimeTaken { get; set; } = string.Empty;
```
Comment (TaoChenOSU) [inline webapi/Models/Response/StepwiseThoughtProcess.cs:22]: nit: specify the unit in the comment since this is a string

File: webapi/Models/Response/BotResponsePrompt.cs (hunk 1 above)
Comment (alliscode) [inline webapi/Models/Response/BotResponsePrompt.cs:15]: Can we pull this out into its own file?

**Review by TaoChenOSU — DISMISSED** [review]

File: webapi/Models/Response/BotResponsePrompt.cs [hunk 3]
```diff
@@ -37,7 +60,7 @@ public class BotResponsePrompt
     /// Relevant additional knowledge extracted using a planner.
     /// </summary>
     [JsonPropertyName("externalInformation")]
-    public string ExternalInformation { get; set; } = string.Empty;
+    public SemanticDependency<StepwiseThoughtProcess> ExternalInformation { get; set; }
```
Comment (alliscode) [inline webapi/Models/Response/BotResponsePrompt.cs:63]: Is BotResponsePrompt currently used for other context types? Will this break them in the time between checking this in and adding the others?

File: webapi/Skills/ChatSkills/ChatSkill.cs [hunk 4]
```diff
@@ -347,6 +348,13 @@ private async Task<ChatMessage> GetChatResponseAsync(string chatId, string userI
         var planResult = await this.AcquireExternalInformationAsync(chatContext, userIntent, externalInformationTokenLimit);
         chatContext.ThrowIfFailed();
 
+        // Extract additional details about planner execution in chat context
+        // TODO: [Issue #150, sk#2106] Accommodate different planner contexts once core team finishes work to return prompt and token usage.
+        var plannerDetails = new SemanticDependency<StepwiseThoughtProcess>(
+                planResult,
+                this._externalInformationSkill.StepwiseThoughtProcess
+            );
```
Comment (alliscode) [inline webapi/Skills/ChatSkills/ChatSkill.cs:356]: Does this work when configured for other planner types?

**Review by alliscode — CHANGES_REQUESTED** [review]

**Review by dehoward — APPROVED** [review]

File: webapi/Models/Response/BotResponsePrompt.cs (hunk 3 above)
Comment (teresaqhoang) [reply webapi/Models/Response/BotResponsePrompt.cs:63]: No, it's only used for the prompt view in the webapp and I've made that backwards compatible with string types too (since we were just returning the raw rendered string to show in the prompt dialog before)

File: webapi/Skills/ChatSkills/ChatSkill.cs (hunk 4 above)
Comment (teresaqhoang) [reply webapi/Skills/ChatSkills/ChatSkill.cs:356]: No, we don't have the ability to get any more context from other planners, but the Core team has a work item to return more context including token usage, so I'll update this when that's done. There's a TODO to track that work.

Until then, context will only exist for Stepwise planner and the prompt dialog will have the same behavior it did before for other planners, which is displaying the proposed plan json.

File: webapi/Models/Response/BotResponsePrompt.cs (hunk 1 above)
Comment (teresaqhoang) [reply webapi/Models/Response/BotResponsePrompt.cs]: Done

File: webapi/Models/Response/StepwiseThoughtProcess.cs (hunk 2 above)
Comment (teresaqhoang) [reply webapi/Models/Response/StepwiseThoughtProcess.cs:23]: Added format definition in comment

**Review by alliscode — APPROVED** [review]

**Review by TaoChenOSU — APPROVED** [review]
